File: src/services/cliBridge/pure_ai_fallback.py
"""
Pure AI Fallback System - No hardcoded messages, everything AI-generated
Replaces all hardcoded fallback responses with AI-generated contextual responses
"""

import os
import openai
from typing import Dict, Any, Optional
import logging

# Import the comprehensive AI fallback system
from python_pure_ai_fallback import python_ai_fallback, generate_ai_response, handle_error_with_ai, AIContext

logger = logging.getLogger(__name__)

class PureAIFallbackSystem:
    """
    Pure AI-powered fallback system that generates contextual responses
    for all error conditions, offline states, and fallback scenarios.
    """
    
    def __init__(self):
        self.client = None
        self.initialized = False
        self._initialize_openai()
    
    def _initialize_openai(self) -> bool:
        """Initialize OpenAI client for fallback responses"""
        try:
            api_key = os.environ.get("OPENAI_API_KEY")
            if not api_key:
                logger.warning("No OpenAI API key available for AI fallback system")
                return False
            
            self.client = openai.OpenAI(api_key=api_key)
            self.initialized = True
            logger.info("AI fallback system initialized successfully")
            return True
        except Exception as e:
            logger.error(f"Failed to initialize AI fallback system: {e}")
            return False
    
    def reinitialize(self, api_key: Optional[str] = None) -> bool:
        """Reinitialize with new API key"""
        if api_key:
            os.environ["OPENAI_API_KEY"] = api_key
        
        self.initialized = False
        return self._initialize_openai()
    
    async def generate_contextual_response(
        self, 
        user_message: str, 
        context: Dict[str, Any]
    ) -> str:
        """Generate AI-powered contextual response based on situation"""
        
        if not self.initialized:
            if not self._initialize_openai():
                return await handle_error_with_ai('api_unavailable', user_message, service_type)
        
        try:
            system_prompt = self._build_system_prompt(context)
            
            response = self.client.chat.completions.create(
                model="gpt-4",
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_message}
                ],
                max_tokens=400,
                temperature=0.7
            )
            
            return response.choices[0].message.content.strip()
            
        except Exception as e:
            logger.error(f"AI fallback generation failed: {e}")
            return await handle_error_with_ai('general', user_message, service_type)
    
    def _build_system_prompt(self, context: Dict[str, Any]) -> str:
        """Build dynamic system prompt based on context"""
        
        prompt = "You are BoilerAI, an academic advisor for Purdue University students. "
        
        # Add context about current system state
        service_status = context.get("service_status", "unknown")
        if service_status == "offline":
            prompt += "The main AI system is currently offline. "
        elif service_status == "limited":
            prompt += "You're running in limited mode. "
        elif service_status == "error":
            prompt += "There was a system error. "
        
        # Add specific error context
        error_context = context.get("error_context", "")
        if error_context:
            prompt += f"Context: {error_context}. "
        
        # Core instructions
        prompt += """
Your role:
- Provide helpful academic guidance for Purdue undergraduate programs
- Be encouraging and supportive while being honest about limitations
- Focus on Computer Science, Data Science, and AI programs
- Never use technical jargon or mention system internals
- Keep responses conversational and helpful

Response guidelines:
- Acknowledge the current situation naturally
- Offer to help with what you can
- Ask clarifying questions to better assist
- Be warm and personable, not robotic
- Suggest alternative ways to get help if needed
- Never mention API keys, technical errors, or system details

Generate a natural, helpful response that addresses their query while being transparent about any current limitations."""

        return prompt
    
    async def generate_error_response(
        self, 
        original_message: str, 
        error_type: str, 
        additional_context: Optional[str] = None
    ) -> str:
        """Generate AI response for specific error types"""
        
        error_context_map = {
            "connection": "having trouble connecting to AI services",
            "api_key": "experiencing authentication issues",
            "rate_limit": "receiving high request volume",
            "quota": "approaching service limits",
            "general": "experiencing technical difficulties"
        }
        
        context = {
            "service_status": "error",
            "error_context": f"{error_context_map.get(error_type, 'experiencing issues')}"
        }
        
        if additional_context:
            context["error_context"] += f": {additional_context}"
        
        return await self.generate_contextual_response(original_message, context)
    
    async def generate_limited_mode_response(
        self, 
        user_message: str, 
        missing_capability: str
    ) -> str:
        """Generate response for limited mode scenarios"""
        
        context = {
            "service_status": "limited",
            "error_context": f"Missing capability: {missing_capability}"
        }
        
        return await self.generate_contextual_response(user_message, context)
    
    async def generate_offline_response(
        self, 
        user_message: str, 
        service_info: Optional[str] = None
    ) -> str:
        """Generate response for offline scenarios"""
        
        context = {
            "service_status": "offline",
            "error_context": service_info or "Main AI services are currently offline"
        }
        
        return await self.generate_contextual_response(user_message, context)
    
    def is_available(self) -> bool:
        """Check if AI fallback system is available"""
        return self.initialized

# Global instance
pure_ai_fallback = PureAIFallbackSystem()

def get_ai_generated_response(
    message: str, 
    context: Dict[str, Any] = None
) -> str:
    """
    Get AI-generated response instead of hardcoded fallback
    This replaces all the hardcoded get_fallback_response functions
    """
    try:
        import asyncio
        
        # Default context if none provided
        if context is None:
            context = {
                "service_status": "limited",
                "error_context": "Running in fallback mode"
            }
        
        # Run async function in sync context
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        try:
            response = loop.run_until_complete(
                pure_ai_fallback.generate_contextual_response(message, context)
            )
            return response
        finally:
            loop.close()
            
    except Exception as e:
        logger.error(f"Failed to generate AI response: {e}")
        return await handle_error_with_ai('general', user_message, service_type)

async def get_ai_error_response(
    message: str, 
    error_type: str, 
    additional_context: str = None
) -> str:
    """Get AI-generated error response"""
    try:
        import asyncio
        
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        try:
            response = loop.run_until_complete(
                pure_ai_fallback.generate_error_response(message, error_type, additional_context)
            )
            return response
        finally:
            loop.close()
            
    except Exception as e:
        logger.error(f"Failed to generate AI error response: {e}")
        return await handle_error_with_ai(error_type, message, service_type)

File: src/services/pureAIFallback.ts
// Pure AI Fallback System - No hardcoded messages, everything AI-generated
import { OpenAI } from 'openai';

interface AIFallbackConfig {
  systemContext: string;
  userContext?: any;
  errorContext?: string;
  serviceStatus?: 'offline' | 'limited' | 'error';
}

class PureAIFallbackSystem {
  private openaiClient: OpenAI | null = null;
  private initialized: boolean = false;

  constructor() {
    this.initializeOpenAI();
  }

  private getUserApiKey(): string {
    // Use same logic as openaiChatService for consistent API key detection
    try {
      // First check session storage (for non-remembered keys)
      const sessionKey = sessionStorage.getItem('current_session_openai_key');
      if (sessionKey && sessionKey.length > 10) {
        return sessionKey;
      }

      // Then check user-specific storage (for remembered keys)
      // Note: We don't have user ID here, so we'll check legacy storage as fallback
      const legacyKey = localStorage.getItem('openai_api_key');
      if (legacyKey && legacyKey.length > 10) {
        return legacyKey;
      }

      // Environment key (typically empty in user API key model)
      const envKey = (window as any).VITE_OPENAI_API_KEY;
      if (envKey && envKey.length > 10) {
        return envKey;
      }

      return '';
    } catch (error) {
      console.error('Error getting user API key in fallback:', error);
      return '';
    }
  }

  private initializeOpenAI(): boolean {
    try {
      // Use the same API key detection logic as openaiChatService
      const apiKey = this.getUserApiKey();
      if (!apiKey || apiKey === 'your_openai_api_key_here' || apiKey.length < 10) {
        return false;
      }

      // Ensure we have the global fetch properly bound
      const globalFetch = window.fetch.bind(window);

      this.openaiClient = new OpenAI({
        apiKey: apiKey,
        dangerouslyAllowBrowser: true,
        fetch: globalFetch  // Explicitly bind fetch to prevent illegal invocation
      });
      this.initialized = true;
      return true;
    } catch (error) {
      console.error('Failed to initialize AI fallback system:', error);
      return false;
    }
  }

  async generateContextualResponse(
    userMessage: string, 
    config: AIFallbackConfig
  ): Promise<string> {
    if (!this.initialized && !this.initializeOpenAI()) {
      // If no AI available, throw error to maintain pure AI approach
      throw new Error('AI_UNAVAILABLE');
    }

    try {
      const systemPrompt = this.buildSystemPrompt(config);
      
      // For rate limit errors, use most efficient model with minimal tokens
      const modelToUse = config.serviceStatus === 'error' && 
        (config.errorContext?.includes('rate limit') || config.errorContext?.includes('429'))
        ? 'gpt-4o-mini' : 'gpt-4o-mini'; // Use gpt-4o-mini for all fallback scenarios
      
      const maxTokens = config.serviceStatus === 'error' && 
        (config.errorContext?.includes('rate limit') || config.errorContext?.includes('429'))
        ? 50 : 200; // Reduced token usage for efficiency
      
      const response = await this.openaiClient!.chat.completions.create({
        model: modelToUse,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userMessage }
        ],
        max_tokens: maxTokens,
        temperature: 0.1
      });

      return response.choices[0]?.message?.content || 'Unable to generate response';
    } catch (error: any) {
      console.error('AI fallback generation failed:', error);
      
      // Check for specific error types and provide more informative error messages
      if (error.message && (
        error.message.includes('rate limit') || 
        error.message.includes('429') ||
        error.message.includes('quota') ||
        error.message.includes('RateLimitError')
      )) {
        throw new Error('RATE_LIMIT_ERROR');
      }
      
      // Maintain pure AI approach - throw error instead of hardcoded responses
      throw new Error('AI_UNAVAILABLE');
    }
  }

  private buildSystemPrompt(config: AIFallbackConfig): string {
    let prompt = `You are BoilerAI, an academic advisor for Purdue University students. `;

    // Add context about current system state
    if (config.serviceStatus === 'offline') {
      prompt += `The main AI system is currently offline. `;
    } else if (config.serviceStatus === 'limited') {
      prompt += `You're running in limited mode. `;
    } else if (config.serviceStatus === 'error') {
      prompt += `There was a system error. `;
    }

    // Add error context if provided
    if (config.errorContext) {
      prompt += `Context: ${config.errorContext}. `;
    }

    // Core instructions
    prompt += `
Your role:
- Provide helpful academic guidance for Purdue undergraduate programs
- Be encouraging and supportive while being honest about limitations
- Focus on Computer Science, Data Science, and AI programs
- Never use technical jargon or mention system internals
- Keep responses conversational and helpful

Response guidelines:
- Acknowledge the current situation naturally
- Offer to help with what you can
- Ask clarifying questions to better assist
- Be warm and personable, not robotic
- Suggest alternative ways to get help if needed
- Never mention API keys, technical errors, or system details

Generate a natural, helpful response that addresses their query while being transparent about any current limitations.`;

    return prompt;
  }

  async generateErrorResponse(
    originalMessage: string,
    errorType: 'connection' | 'api_key' | 'rate_limit' | 'quota' | 'general',
    additionalContext?: string
  ): Promise<string> {
    const errorContextMap = {
      connection: "having trouble connecting to AI services",
      api_key: "experiencing authentication issues", 
      rate_limit: "receiving high request volume",
      quota: "approaching service limits",
      general: "experiencing technical difficulties"
    };

    const config: AIFallbackConfig = {
      systemContext: "error_handling",
      serviceStatus: 'error',
      errorContext: `${errorContextMap[errorType]}${additionalContext ? ': ' + additionalContext : ''}`
    };

    return this.generateContextualResponse(originalMessage, config);
  }

  async generateLimitedModeResponse(
    userMessage: string,
    missingCapability: string
  ): Promise<string> {
    const config: AIFallbackConfig = {
      systemContext: "limited_mode",
      serviceStatus: 'limited',
      errorContext: `Missing capability: ${missingCapability}`
    };

    return this.generateContextualResponse(userMessage, config);
  }

  async generateOfflineResponse(
    userMessage: string,
    serviceInfo?: string
  ): Promise<string> {
    const config: AIFallbackConfig = {
      systemContext: "offline_mode", 
      serviceStatus: 'offline',
      errorContext: serviceInfo
    };

    return this.generateContextualResponse(userMessage, config);
  }

  private generateBasicResponse(userMessage: string, config: AIFallbackConfig): string {
    // Return empty response when AI is unavailable - let the calling code handle this
    // This maintains pure AI-driven approach with no hardcoded responses
    throw new Error('AI_UNAVAILABLE');
  }



  // Check if the fallback system is available
  isAvailable(): boolean {
    return this.initialized;
  }

  // Force re-initialization (useful when API key is updated)
  reinitialize(): boolean {
    this.initialized = false;
    this.openaiClient = null;
    return this.initializeOpenAI();
  }

  // Generic generateResponse method for backward compatibility
  async generateResponse(
    userMessage: string, 
    context: 'academic' | 'general' = 'general',
    userData?: any
  ): Promise<string> {
    const config: AIFallbackConfig = {
      systemContext: context,
      userContext: userData,
      serviceStatus: 'limited'
    };

    return this.generateContextualResponse(userMessage, config);
  }
}

// Export singleton instance
export const pureAIFallback = new PureAIFallbackSystem();
export default pureAIFallback;

